lib/uktest/test.c

```diff
@@ -0,0 +1,224 @@
+/*
+ * Authors: Alexander Jung <a.jung@lancs.ac.uk>
+ *          Marc Rittinghaus <marc.rittinghaus@kit.edu>
+ *
+ * Copyright (c) 2021, Lancaster University.  All rights reserved.
+ * Copyright (c) 2021, Karlsruhe Institute of Technology (KIT).
+ *                     All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <limits.h>
+
+#include <uk/test.h>
+#include <uk/list.h>
+#include <uk/config.h>
+#include <uk/essentials.h>
+
+#ifdef CONFIG_LIBUKTEST_LOG_STATS
+/**
+ * Test suite, case and assertion statistics
+ */
+struct uk_test_stats {
+	unsigned int total;
+	unsigned int fail;
+	unsigned int success;
+};
+
+static struct uk_test_stats testsuite_stats = {0};
+static struct uk_test_stats testcase_stats = {0};
+static struct uk_test_stats testassert_stats = {0};
+
+#include <uk/plat/console.h>
+
+#define UK_TEST_STATS_INIT_CLASS UK_INIT_CLASS_LATE
+#define UK_TEST_STATS_INIT_PRIO  9 /* As late as possible. */
+
+#if CONFIG_LIBUKDEBUG_ANSI_COLOR
+#define UK_TEST_STAT_FAILED	UK_ANSI_MOD_BOLD \
+				UK_ANSI_MOD_COLORFG(UK_ANSI_COLOR_RED)
+#else /* CONFIG_LIBUKDEBUG_ANSI_COLOR */
+#define UK_TEST_STAT_FAILED
+#endif /* !CONFIG_LIBUKDEBUG_ANSI_COLOR */
+
+static int
+uk_test_print_stats(void)
+{
+	int failed;
+
+	/*
+	 * Test suites
+	 */
+
+	uk_pr_info("uktest:suites:     ");
+	failed = testsuite_stats.fail;
+	if (failed > 0)
+		uk_pr_info(UK_TEST_STAT_FAILED
+			  "%d failed"
```

Comment:
Just a nitpick while you edit the PR anyways. One space missing for proper indentation 😉 Also with the following `uk_pr_info`.